src/easy/handler.rs

```diff
@@ -1924,10 +1924,13 @@ impl<H> Easy2<H> {
     /// By default this option is not set and corresponds to
     /// `CURLOPT_TIMEOUT_MS`.
     pub fn timeout(&mut self, timeout: Duration) -> Result<(), Error> {
-        // TODO: checked arithmetic and casts
-        // TODO: use CURLOPT_TIMEOUT if the timeout is too great
-        let ms = timeout.as_secs() * 1000 + timeout.subsec_millis() as u64;
-        self.setopt_long(curl_sys::CURLOPT_TIMEOUT_MS, ms as c_long)
+        let mut ms = timeout.as_secs() * 1000 + timeout.subsec_millis() as u64;
+        if ms > std::i32::MAX as u64 {
+            ms = ms / 1000;
+            self.setopt_long(curl_sys::CURLOPT_TIMEOUT, ms as c_long)
+        } else {
+            self.setopt_long(curl_sys::CURLOPT_TIMEOUT_MS, ms as c_long)
+        }
```

Comment:
Could this perhaps be modeled as:

```rust
match c_long::try_from(ms) {
    Ok(amt) => self.setopt_long(curl_sys::CURLOPT_TIMEOUT_MS, amt),
    Err(_) => {
        let amt = c_long::try_from(ms / 1000).unwrap();
        self.setopt_long(curl_sys::CURLOPT_TIMEOUT, amt)
    }
}
```

that avoids baking in knowledge about `i32` and `c_long`. Additionally it avoid confusingly storing a seconds-based value in a variable naed `ms`. Finally it asserts that the seconds conversion is indeed lossless.